File: algorithm_data_structure/leetcode/695.max-area-of-island.py
#
# @lc app=leetcode id=695 lang=python3
#
# [695] Max Area of Island
#

# @lc code=start
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        # Main function: iterate over all possible points.
        ans = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    grid[i][j] = 0  # Mark this point as already searched.
                    ans = max(self.dfs(grid, i, j), ans)
                    # ans = max(self.bfs(grid, i, j), ans)
        return ans

    def dfs(self, grid, i, j):
        # DFS based on stack
        stack = [(i, j)]
        area = 0
        # Stack for DFS
        while stack:
            r, c = stack.pop(-1)
            area += 1
            for nr, nc in ((r - 1, c), (r + 1, c), (r, c - 1), (r, c + 1)):
                if 0 <= nr < len(grid) and 0 <= nc < len(grid[0]) and grid[nr][nc] == 1:
                    stack.append((nr, nc))
                    grid[nr][nc] = 0
        return area


# @lc code=end

File: algorithm_data_structure/leetcode/695.max-area-of-island.cpp
/*
 * @lc app=leetcode id=695 lang=cpp
 *
 * [695] Max Area of Island
 */

// @lc code=start

#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;
class Solution {
   public:
	// {-1, 0} -> left, {0, 1} -> up
	// {+1, 0} -> right, {0, -1} -> down
	vector<int> directions{-1, 0, 1, 0, -1};

	int dfs(vector<vector<int>> &grid, int row, int col) {
		// Check base case, namely when to return the value
		if (grid[row][col] == 0) {
			return 0;
		}
		// Mark this land as already searched
		grid[row][col] = 0;

		int x = 0, y = 0, area = 1;	 // The area == 1, so means that this island
		// Search over different directions
		for (int i = 0; i < 4; i++) {
			x = row + directions[i];
			y = col + directions[i + 1];
			// Firstly check the boundary before recursion
			if (x >= 0 && x < grid.size() && y >= 0 && y < grid[0].size()) {
				// DFS recursive searching -> Like a stack.
				area = area + dfs(grid, x, y);
			}
		}

		return area;
	}

	int maxAreaOfIsland(vector<vector<int>> &grid) {
		// Check base case
		if (grid.empty() || grid[0].empty()) {
			return 0;
		}

		int max_area = 0;

		// Iterate over the map with DFS searching
		for (int i = 0; i < grid.size(); i++) {
			for (int j = 0; j < grid[0].size(); j++) {
				if (grid[i][j] == 1) {
					max_area = max(max_area, dfs(grid, i, j));
				}
			}
		}

		return max_area;
	}
};
// @lc code=end
